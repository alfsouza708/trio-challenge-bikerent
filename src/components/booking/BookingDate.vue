<script lang="ts">
import { defineComponent } from 'vue'
import VueDatePicker from '@vuepic/vue-datepicker'
import moment from 'moment'
import '@vuepic/vue-datepicker/dist/main.css'

export default defineComponent({
  name: 'BookingDate',
  components: { VueDatePicker },
  emits: ['update:date'],
  data() {
    return {
      date: null,
      selecting: false,
      windowWidth: window.innerWidth,
      windowHeight: window.innerHeight * 0.8
    }
  },
  computed: {
    minDate() {
      return moment().toDate()
    },
    maxDate() {
      return moment().add(3, 'M').toDate()
    },
    isMobile() {
      return this.windowWidth <= 475
    }
  },
  watch: {
    selecting() {
      this.date && !this.selecting && !this.isMobile
        ? this.$emit('update:date', { isSelected: true, dateToRent: this.formatDates(this.date) })
        : this.$emit('update:date', { isSelected: false, dateToRent: [] })
    }
  },
  mounted() {
    window.addEventListener('resize', () => {
      this.windowWidth = window.innerWidth
    })
  },
  methods: {
    handleModelUpdate() {
      if (this.isMobile && this.date) {
        this.$emit('update:date', { isSelected: true, dateToRent: this.formatDates(this.date) })
      }
    },
    formatDates(date: Date[]) {
      if (date) {
        return date.map((date) => moment(date).format('YYYY-MM-DD'))
      }
    },
    formatDatesMobile(date: Date[]) {
      const datesFormated = date.map((date) => moment(date).format('MMM/DD'))

      return `From ${datesFormated[0]} to ${datesFormated[1]}`
    }
  }
})
</script>

<template>
  <section class="booking-date">
    <vue-date-picker
      v-model="date"
      uid="date-selector"
      :inline="!isMobile"
      :auto-apply="!isMobile"
      :teleport="true"
      range
      month-name-format="long"
      :hide-navigation="['time']"
      :min-date="minDate"
      :max-date="maxDate"
      prevent-min-max-navigation
      input-class-name="dp-custom-input"
      :clearable="false"
      :format="formatDatesMobile"
      :action-row="{ showCancel: false }"
      :config="{
        modeHeight: windowHeight,
        onClickOutside: () => {
          isMobile ? true : false
        }
      }"
      @range-start="() => (selecting = true)"
      @range-end="
        () => {
          if (selecting && !isMobile) selecting = false
        }
      "
      @update:model-value="handleModelUpdate"
    />
  </section>
</template>

<style lang="scss">
:root {
  --dp-font-family: 'Mont';
  --dp-border-radius: 40px;
  --dp-cell-size: 40px;
  --dp-menu-min-width: 375px;
  --dp-cell-padding: 12px;
  --dp-menu-padding: 24px 26px;
}

.dp__theme_light {
  --dp-background-color: #1f49d1;
  --dp-text-color: #ffffff;
  --dp-disabled-color-text: rgba(255, 255, 255, 0.5);
  --dp-icon-color: #1f49d1;
  --dp-primary-color: #577cdd;
  --dp-border-color-hover: #ffffff;
  --dp-hover-icon-color: #1f49d1;
  --dp-hover-text-color: #1f49d1;
  --dp-range-between-dates-background-color: var(--dp-hover-color, #c1cff2);
  --dp-disabled-color: rgba(0, 0, 0, 0);
}

.dp-custom-input {
  color: black;
  background-color: #fff;
}

.dp__action_row {
  justify-content: center;

  .dp__action_buttons {
    margin: 0;
  }

  .dp__action_select {
    display: flex;
    justify-content: center;
    align-items: center;
    background: #ffd775;
    color: black;
    font-size: 1rem;
    font-weight: bold;
    border-radius: 20px;
    margin: 0;
    width: 327px;
    height: 60px;
  }

  .dp__selection_preview {
    display: none;
  }
}

.dp__action_buttons {
  margin: 0;
}

.dp__action_select {
  display: flex;
  justify-content: center;
  align-items: center;
  background: #ffd775;
  color: black;
  font-size: 1rem;
  font-weight: bold;
  margin: 0;
  width: 327px;
  height: 60px;
}

.dp__selection_preview {
  display: none;
}

.dp__month_year_row {
  height: 5rem;

  .dp__overlay {
    border-radius: 1.5rem;
  }

  .dp--arrow-btn-nav {
    display: flex;
    justify-content: center;
    align-items: center;

    .dp__inner_nav {
      color: #fff;
      border: #fff solid 1px;
      border-radius: 20px;
      padding: 4px;
      width: 52px;
      height: 52px;
      background: none;

      &.dp__inner_nav_disabled {
        color: rgba(255, 255, 255, 0.5);
        border: rgba(255, 255, 255, 0.5) solid 1px;
      }
    }
  }

  button[aria-label='Previous month'] {
    order: 1;
  }
  button[aria-label='Next month'] {
    order: 2;
  }

  .dp__month_year_wrap {
    flex-direction: column;
    pointer-events: none;

    button[aria-label='Open months overlay'] {
      font-size: 34px;
      font-weight: bold;
      justify-content: flex-start;
      color: rgba(255, 255, 255, 1);
    }

    button[aria-label='Open years overlay'] {
      height: fit-content;
      width: fit-content;
      padding: 8px;
      font-size: medium;
      font-weight: normal;
      color: rgba(255, 255, 255, 0.5);
    }
  }
}

.dp__calendar {
  .dp__calendar_header {
    font-size: 16px;
    font-weight: normal;
    color: rgba(255, 255, 255, 0.5);
  }

  .dp__calendar_header_separator {
    display: none;
  }

  .dp__calendar .dp__calendar_row .dp__calendar_item {
    .dp__today {
      border-radius: 50%;
      border-color: #fff;
    }
    .dp__range_start {
      color: #1f49d1;
      border-radius: 50%;
      background-color: #fff;
      overflow: visible;
      z-index: 2;

      &::after {
        content: '';
        position: absolute;
        width: 24px;
        height: 40px;
        right: -8px;
        background-color: #577cdd;
        z-index: 0;
        mix-blend-mode: color;
      }
    }

    .dp__range_between {
      color: #fff;
      border: none;
      background-color: #577cdd;
      overflow: visible;
      z-index: 2;

      &::after {
        content: '';
        position: absolute;
        width: 50px;
        height: 40px;
        left: 0;
        background-color: #577cdd;
        z-index: 0;
        mix-blend-mode: color;
      }
    }

    .dp__range_end {
      color: #1f49d1;
      border-radius: 50%;
      background-color: #fff;
      overflow: visible;
      z-index: 2;

      &::before {
        content: '';
        position: absolute;
        width: 24px;
        height: 40px;
        left: -6px;
        background-color: #577cdd;
        z-index: 0;
        mix-blend-mode: color;
      }
    }

    .dp__range_start.dp__range_end {
      &::before,
      &::after {
        display: none;
      }
    }
  }
}
</style>
