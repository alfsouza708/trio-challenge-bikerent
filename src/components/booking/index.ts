export { default as BookingAddressMap } from './BookingAddressMap.vue'
export { default as BookingPricing } from './BookingPricing.vue'
export { default as BookingDate } from './BookingDate.vue'
export { default as BookingFeedback } from './BookingFeedback.vue'
