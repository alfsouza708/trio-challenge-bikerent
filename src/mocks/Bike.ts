import type { BikeRent, BikeRentDetails, BikeReturnDetails } from '@/core/api/modules/bike'

const [mockDate] = new Date().toISOString().split('T')

const mockedImageUrls = [
  'https://cremecycles.com/images/glowne/15.jpg',
  'https://cremecycles.com/images/glowne/13.jpg',
  'https://cremecycles.com/images/glowne/14.jpg'
]

const mockedBike: Bike = {
  id: 1,
  candidateId: 5,
  name: 'Murazik, Thiel and Robel',
  type: 'Cyclocross Bicycle',
  description:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
  bodySize: 24,
  city: 'Manhattan',
  cardImage: 'https://cremecycles.com/images/glowne/15.jpg',
  imageUrls: mockedImageUrls,
  maxLoad: 110,
  rate: 135,
  ratings: 4.8
}

const mockedPricingDetails: BikeRent = {
  fee: 10,
  rentAmount: 120,
  totalAmount: 130
}

const mockBikeRentDetails: BikeRentDetails = {
  userId: 1,
  bikeId: mockedBike.id,
  dateFrom: mockDate,
  dateTo: mockDate
}

const mockedBikeReturnDetails: BikeReturnDetails = {
  userId: 1,
  bikeId: mockedBike.id
}

const mockedBikeList = [mockedBike, mockedBike, mockedBike, mockedBike]

const mockedAddress = '745 Atlantic Ave, Boston, MA 02111, United States'

export {
  mockedBike,
  mockedBikeList,
  mockedImageUrls,
  mockBikeRentDetails,
  mockedBikeReturnDetails,
  mockedAddress,
  mockedPricingDetails
}
