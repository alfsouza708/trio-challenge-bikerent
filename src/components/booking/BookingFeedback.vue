<script lang="ts">
import { defineComponent } from 'vue'

import { Chip } from '@/components/chip'
import { ImageLazy } from '@/components/image'

export default defineComponent({
  name: 'BookingFeedback',
  components: { Chip, ImageLazy },
  props: {
    imageToRender: {
      type: String,
      required: true
    },
    bikeName: {
      type: String,
      required: true
    },
    chipText: {
      type: String,
      required: true
    }
  },
  data() {
    return {}
  },
  computed: {}
})
</script>

<template>
  <section class="booking-feedback card">
    <h2 class="font-bold text-2xl">Thank you!</h2>
    <p class="font-normal text-md">Your bike is booked.</p>
    <image-lazy :src="imageToRender" class="image--thanks-card" />
    <p class="font-semibold text-md">{{ bikeName }}</p>
    <chip color="secondary" size="sm">{{ chipText }}</chip>
  </section>
</template>

<style lang="scss">
.booking-feedback {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  padding: 40px;

  .image--thanks-card {
    width: 185px;
  }
}
</style>
