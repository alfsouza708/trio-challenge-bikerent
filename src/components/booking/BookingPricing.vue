<script lang="ts">
import type { BikeRent } from '@/core/api/modules/typings/bike'
import { defineComponent, type PropType } from 'vue'
import { Icon } from '@/components/icon/'
import type { CurrencyCode } from '@/core/config'
import { formatCurrency } from '@/core/helpers/currency'

export default defineComponent({
  name: 'BookingPricing',
  components: { Icon },
  props: {
    rentAmountDetails: {
      type: Object as PropType<BikeRent>,
      required: true
    },
    currency: {
      type: String as PropType<CurrencyCode>,
      required: true
    }
  },
  computed: {
    rentAmount(): number {
      return this.rentAmountDetails.rentAmount
    },
    fee(): number {
      return this.rentAmountDetails.fee
    },
    totalAmount(): number {
      return this.rentAmountDetails.totalAmount
    }
  },
  methods: {
    formatPrice(value: number): String {
      return formatCurrency(value, this.currency)
    }
  }
})
</script>

<template>
  <section class="booking-pricing">
    <div class="grid grid-cols-2 gap-y-3 text-sm">
      <div class="flex items-center">
        Subtotal
        <icon type="solid" size="xs" class="ml-1">circle-info</icon>
      </div>
      <div class="text-right">{{ formatPrice(rentAmount) }}</div>

      <div class="flex items-center">
        Service Fee
        <icon type="solid" size="xs" class="ml-1">circle-info</icon>
      </div>
      <div class="text-right">{{ formatPrice(fee) }}</div>

      <div class="flex items-center text-base">Total</div>
      <div class="text-right text-2xl font-bold">{{ formatPrice(totalAmount) }}</div>
    </div>
  </section>
</template>

<style lang="scss" scoped>
.booking-pricing {
  position: relative;
}
</style>
