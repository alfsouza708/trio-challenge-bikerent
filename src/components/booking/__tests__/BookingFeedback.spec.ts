import { beforeEach, describe, expect, it, vi } from 'vitest'

import { BookingFeedback } from '@/components/booking'
import { mockedBike } from '@/mocks'
import { mockIntersectionObserver } from '@/core/helpers/tests'

import { mount } from '@vue/test-utils'

describe('BookingFeedback', () => {
  vi.stubGlobal('IntersectionObserver', mockIntersectionObserver)

  beforeEach(() => {
    vi.clearAllMocks()
  })

  it('renders correctly', () => {
    const wrapper = mount(BookingFeedback, {
      props: { imageToRender: mockedBike.cardImage as string, bikeName: mockedBike.name, chipText: mockedBike.type }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('prop imageToRender', () => {
    const wrapper = mount(BookingFeedback, {
      props: { imageToRender: mockedBike.cardImage as string, bikeName: mockedBike.name, chipText: mockedBike.type }
    })

    expect(wrapper.findComponent('image-lazy'))
  })

  it('prop chipText', () => {
    const wrapper = mount(BookingFeedback, {
      props: { imageToRender: mockedBike.cardImage as string, bikeName: mockedBike.name, chipText: mockedBike.type }
    })

    expect(wrapper.findComponent('chip'))
  })
})
