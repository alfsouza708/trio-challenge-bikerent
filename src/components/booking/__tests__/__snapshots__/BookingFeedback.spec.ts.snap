// Vitest Snapshot v1

exports[`BookingFeedback > renders correctly 1`] = `
"<section class=\\"booking-feedback card\\">
  <h2 class=\\"font-bold text-2xl\\">Thank you!</h2>
  <p class=\\"font-normal text-md\\">Your bike is booked.</p><img class=\\"image image--lazy image--thanks-card\\" data-v-ff3f7011=\\"\\">
  <p class=\\"font-semibold text-md\\">Murazik, Thiel and Robel</p>
  <div class=\\"chip chip--sm chip--secondary\\" data-v-221b4c58=\\"\\">
    <div class=\\"chip__content\\" data-v-221b4c58=\\"\\">Cyclocross Bicycle</div>
  </div>
</section>"
`;
